File: go/030.go
func findSubstring(s string, words []string) []int {
    size, wordCnt, res := len(s), len(words), []int{}
    if size == 0 || wordCnt == 0 {
        return res
    }
    dict, wz := map[string]int{}, len(words[0])
    for _, word := range words {
        dict[word]++
    }
    for i := 0; i < wz; i++ {
        left, cnt, tempDict := i, 0, map[string]int{}
        for j := left; j <= size-wz; j += wz {
            cur := s[j : j+wz]
            if _, ok := dict[cur]; !ok {
                tempDict, cnt, left = map[string]int{}, 0, j+wz
                continue
            }
            tempDict[cur]++
            for tempDict[cur] > dict[cur] {
                t := s[left : left+wz]
                tempDict[t]--
                cnt--
                left += wz
            }
            cnt++
            if cnt == wordCnt {
                res = append(res, left)
                t := s[left : left+wz]
                left += wz
                tempDict[t]--
                cnt--
            }
        }
    }
    return res
}

File: cpp/030.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> ret;
        if (s.empty() || words.empty())
            return ret;
        int size = s.size(), wz = words[0].size(), cnt = words.size();
        map<string, int> dict;
        for (auto str : words)
            ++dict[str];
        for (int i = 0; i < wz; ++i) {
            int left = i, matchCnt = 0;
            map<string, int> temp;
            for (int j = i; j <= size - wz; j += wz) {
                string cur = s.substr(j, wz);
                if (dict.find(cur) == dict.end()) {
                    temp.clear();
                    left = j + wz;
                    matchCnt = 0;
                    continue;
                }
                ++temp[cur];
                while (temp[cur] > dict[cur]) {
                    string t = s.substr(left, wz);
                    --temp[t];
                    --matchCnt;
                    left += wz;
                }
                if (++matchCnt == cnt) {
                    ret.push_back(left);
                    string t = s.substr(left, wz);
                    --temp[t];
                    --matchCnt;
                    left += wz;
                }
            }
        }
        return ret;
    }
};
